app/dl/iter.go

```diff
@@ -168,13 +185,142 @@ func (i *iter) process(ctx context.Context) (ret bool, skip bool) {
+	// 1. --skip-same flag is set
+	// 2. --force-web-check is NOT set
+	// 3. Either:
+	//    a) JSON export includes raw Telegram data (--raw flag, works with any template), OR
+	//    b) Using default template with standard JSON export
+	if i.opts.SkipSame && !i.opts.ForceWebCheck {
+		dialog := i.dialogs[i.dialogIndex]
+		const defaultTemplate = `{{ .DialogID }}_{{ .MessageID }}_{{ filenamify .FileName }}`
+		isDefaultTemplate := i.opts.Template == defaultTemplate
+		canOptimize := dialog.HasRawData || isDefaultTemplate
+
+		// Log optimization status on first message
+		if i.logicalPos == 0 {
+			// Warn if using custom template without MessageID (collision risk)
+			if canOptimize && !strings.Contains(i.opts.Template, "MessageID") {
+				logctx.From(ctx).Warn("Template does not include MessageID - filename collisions may occur",
+					zap.String("current_template", i.opts.Template),
+					zap.String("recommendation", "Include {{ .MessageID }} in template to ensure unique filenames"),
+					zap.String("note", "Files with duplicate names will be skipped by --skip-same"))
+			}
+
+			if !canOptimize {
+				logctx.From(ctx).Warn("Skip-same optimization disabled",
+					zap.String("reason", "requires either raw JSON export or default template"),
+					zap.String("solution", "Use --raw flag during export, or use default template"),
+					zap.Bool("has_raw_data", dialog.HasRawData),
+					zap.Bool("is_default_template", isDefaultTemplate),
+					zap.String("note", "Falling back to network-based file checking"))
+			} else if len(dialog.MessageMetas) > 0 {
+				logctx.From(ctx).Info("Skip-same optimization enabled",
+					zap.Int("messages_with_metadata", len(dialog.MessageMetas)),
+					zap.Int("total_messages", len(dialog.Messages)),
+					zap.Bool("has_raw_data", dialog.HasRawData),
+					zap.Bool("is_default_template", isDefaultTemplate),
+					zap.String("note", "Using JSON metadata to skip files without network calls. Use --force-web-check to disable."))
+			} else {
+				logctx.From(ctx).Warn("Skip-same optimization unavailable",
+					zap.String("reason", "no metadata in JSON export"),
+					zap.String("note", "Files will require network calls to check"))
+			}
+		}
+
+		// Only proceed with optimization if requirements are met
+		if canOptimize && len(dialog.MessageMetas) > 0 {
+			if meta, ok := dialog.MessageMetas[msg]; ok && meta.Filename != "" {
+				// Extract peer ID from InputPeerClass without network call
+				var peerID int64
+				switch p := dialog.Peer.(type) {
+				case *tg.InputPeerChannel:
+					peerID = p.ChannelID
+				case *tg.InputPeerUser:
+					peerID = p.UserID
+				case *tg.InputPeerChat:
+					peerID = p.ChatID
+				default:
+					// Unknown peer type, skip optimization for this message
+					if i.logicalPos <= 3 {
+						logctx.From(ctx).Debug("Quick skip-same: unknown peer type, using network check",
+							zap.String("peer_type", fmt.Sprintf("%T", dialog.Peer)))
+					}
+					goto skipOptimization
```

Comment:
Plz avoid using `goto` as much as possible, try to extract related code to a function. Otherwise, please provide sufficiently reasons.